File: packages/ensemble/src/weights.js
import { ValidationError } from '@wlearn/core'

/**
 * Project vector onto the probability simplex {w: w >= 0, sum(w) = 1}.
 * O(n log n) algorithm from Duchi et al. 2008.
 *
 * @param {Float64Array} v - input vector
 * @returns {Float64Array} - projected vector
 */
export function projectSimplex(v) {
  const n = v.length
  if (n === 0) return new Float64Array(0)
  if (n === 1) return new Float64Array([1.0])

  // Sort descending
  const u = new Float64Array(v)
  u.sort()
  u.reverse()

  const cssv = new Float64Array(n)
  cssv[0] = u[0]
  for (let i = 1; i < n; i++) {
    cssv[i] = cssv[i - 1] + u[i]
  }

  let rho = -1
  for (let i = 0; i < n; i++) {
    if (u[i] * (i + 1) > cssv[i] - 1) {
      rho = i
    }
  }

  if (rho < 0) {
    // Fallback: uniform
    const out = new Float64Array(n)
    out.fill(1.0 / n)
    return out
  }

  const theta = (cssv[rho] - 1.0) / (rho + 1.0)
  const out = new Float64Array(n)
  for (let i = 0; i < n; i++) {
    out[i] = Math.max(v[i] - theta, 0.0)
  }
  return out
}

/**
 * Optimize ensemble weights via projected gradient descent on the simplex.
 *
 * Classification: minimizes negative log-loss.
 * Regression: minimizes MSE.
 *
 * @param {Float64Array[]} oofPredictions - per-model OOF predictions
 * @param {TypedArray} yTrue - true labels
 * @param {Float64Array} initWeights - initial weights (e.g. from Caruana)
 * @param {object} opts
 * @returns {Float64Array} - optimized weights (>= 0, sum = 1)
 */
export function optimizeWeights(oofPredictions, yTrue, initWeights, {
  task = 'classification',
  lr = 0.05,
  nIter = 100,
} = {}) {
  const nModels = oofPredictions.length
  const n = yTrue.length

  if (nModels === 0) {
    throw new ValidationError('optimizeWeights: need at least 1 model')
  }
  if (nModels === 1) {
    return new Float64Array([1.0])
  }

  const w = new Float64Array(initWeights)
  const eps = 1e-15

  if (task === 'classification') {
    const predLen = oofPredictions[0].length
    const nc = predLen / n
    if (nc !== Math.floor(nc)) {
      throw new ValidationError('optimizeWeights: prediction length must be divisible by n')
    }

    for (let iter = 0; iter < nIter; iter++) {
      const grad = new Float64Array(nModels)

      for (let i = 0; i < n; i++) {
        const c = yTrue[i] | 0
        // Ensemble probability for true class
        let pTrue = 0
        for (let m = 0; m < nModels; m++) {
          pTrue += w[m] * oofPredictions[m][i * nc + c]
        }
        pTrue = Math.max(pTrue, eps)

        for (let m = 0; m < nModels; m++) {
          grad[m] -= oofPredictions[m][i * nc + c] / pTrue
        }
      }

      // Normalize gradient
      for (let m = 0; m < nModels; m++) {
        grad[m] /= n
      }

      // Gradient step + project
      for (let m = 0; m < nModels; m++) {
        w[m] -= lr * grad[m]
      }
      const proj = projectSimplex(w)
      for (let m = 0; m < nModels; m++) w[m] = proj[m]
    }
  } else {
    // Regression: minimize MSE
    for (let iter = 0; iter < nIter; iter++) {
      const grad = new Float64Array(nModels)

      for (let i = 0; i < n; i++) {
        let pred = 0
        for (let m = 0; m < nModels; m++) {
          pred += w[m] * oofPredictions[m][i]
        }
        const residual = yTrue[i] - pred
        for (let m = 0; m < nModels; m++) {
          grad[m] -= 2 * residual * oofPredictions[m][i]
        }
      }

      for (let m = 0; m < nModels; m++) {
        grad[m] /= n
      }

      for (let m = 0; m < nModels; m++) {
        w[m] -= lr * grad[m]
      }
      const proj = projectSimplex(w)
      for (let m = 0; m < nModels; m++) w[m] = proj[m]
    }
  }

  return w
}

File: py/wlearn/ensemble/_weights.py
"""Ensemble weight optimization via projected gradient descent on the simplex.

Refines frequency-based Caruana weights to optimized weights using gradient
descent with simplex projection (Duchi et al. 2008).
"""

import numpy as np


def project_simplex(v):
    """Project vector v onto the probability simplex {w: w >= 0, sum(w) = 1}.

    Algorithm from Duchi et al. (2008), O(n log n).

    Args:
        v: np.ndarray of shape (n,)

    Returns:
        np.ndarray of shape (n,) on the simplex
    """
    n = len(v)
    if n == 0:
        return v.copy()
    if n == 1:
        return np.array([1.0])

    u = np.sort(v)[::-1]
    cssv = np.cumsum(u)
    rho_candidates = np.nonzero(u * np.arange(1, n + 1) > (cssv - 1))[0]
    if len(rho_candidates) == 0:
        # Fallback: uniform weights
        return np.full(n, 1.0 / n)
    rho = rho_candidates[-1]
    theta = (cssv[rho] - 1.0) / (rho + 1.0)
    return np.maximum(v - theta, 0.0)


def optimize_weights(oof_predictions, y_true, init_weights,
                     task='classification', lr=0.05, n_iter=100):
    """Optimize ensemble member weights via projected gradient descent.

    For classification: minimizes negative log-loss over OOF probabilities.
    For regression: minimizes MSE over OOF predictions.
    Constraints: weights >= 0, sum(weights) = 1.

    Args:
        oof_predictions: list of np.ndarray (one per selected model)
            Classification: each (n * n_classes,) flat row-major
            Regression: each (n,)
        y_true: np.ndarray of true labels
        init_weights: np.ndarray of initial weights (from Caruana selection)
        task: 'classification' or 'regression'
        lr: learning rate for gradient descent
        n_iter: number of gradient steps

    Returns:
        np.ndarray of optimized weights (>= 0, sum = 1)
    """
    n = len(y_true)
    m = len(oof_predictions)

    if m == 0:
        return np.array([], dtype=np.float64)
    if m == 1:
        return np.array([1.0])

    w = init_weights.copy().astype(np.float64)
    # Ensure starting point is on simplex
    w = project_simplex(w)

    eps = 1e-15

    if task == 'classification':
        n_classes = len(oof_predictions[0]) // n
        # Precompute: for each model m, extract proba[i, y_true[i]]
        # and full proba matrix for gradient computation
        y_int = np.asarray(y_true, dtype=np.int32)

        for _ in range(n_iter):
            # Compute ensemble proba for true class: p_i = sum(w_m * oof_m[i*nc + y_i])
            p_true = np.zeros(n, dtype=np.float64)
            for j in range(m):
                wj = w[j]
                oof = oof_predictions[j]
                for i in range(n):
                    p_true[i] += wj * oof[i * n_classes + y_int[i]]

            # Clip for numerical stability
            np.maximum(p_true, eps, out=p_true)

            # Gradient: d(loss)/d(w_m) = -1/n * sum(oof_m[i*nc+y_i] / p_true[i])
            grad = np.zeros(m, dtype=np.float64)
            for j in range(m):
                oof = oof_predictions[j]
                g = 0.0
                for i in range(n):
                    g += oof[i * n_classes + y_int[i]] / p_true[i]
                grad[j] = -g / n

            w = project_simplex(w - lr * grad)

    else:
        # Regression: minimize MSE
        for _ in range(n_iter):
            # Compute ensemble prediction: p_i = sum(w_m * oof_m[i])
            p = np.zeros(n, dtype=np.float64)
            for j in range(m):
                wj = w[j]
                oof = oof_predictions[j]
                for i in range(n):
                    p[i] += wj * float(oof[i])

            # Residuals
            resid = np.asarray(y_true, dtype=np.float64) - p

            # Gradient: d(MSE)/d(w_m) = -2/n * sum(resid[i] * oof_m[i])
            grad = np.zeros(m, dtype=np.float64)
            for j in range(m):
                oof = oof_predictions[j]
                g = 0.0
                for i in range(n):
                    g += resid[i] * float(oof[i])
                grad[j] = -2.0 * g / n

            w = project_simplex(w - lr * grad)

    return w
